core/services/job_runner.go

```diff
@@ -1,276 +1,83 @@
 package services
 
 import (
-	"errors"
 	"fmt"
 	"sync"
 
-	"github.com/smartcontractkit/chainlink/core/adapters"
 	"github.com/smartcontractkit/chainlink/core/logger"
-	"github.com/smartcontractkit/chainlink/core/store"
 	"github.com/smartcontractkit/chainlink/core/store/models"
-	"go.uber.org/multierr"
 )
 
+//go:generate mockery -name JobRunner -output ../internal/mocks/ -case=underscore
+
 // JobRunner safely handles coordinating job runs.
 type JobRunner interface {
 	Start() error
 	Stop()
-	resumeRunsSinceLastShutdown() error
-	channelForRun(*models.ID) chan<- struct{}
-	workerCount() int
+	Run(*models.JobRun) error
+
+	WorkerCount() int
 }
 
 type jobRunner struct {
-	started              bool
-	done                 chan struct{}
-	bootMutex            sync.Mutex
-	store                *store.Store
-	workerMutex          sync.RWMutex
-	workers              map[string]chan struct{}
-	workersWg            sync.WaitGroup
-	demultiplexStopperWg sync.WaitGroup
+	workersMutex sync.RWMutex
+	workers      map[string]struct{}
+	workersWg    sync.WaitGroup
+
+	jobExecutor JobExecutor
 }
 
 // NewJobRunner initializes a JobRunner.
-func NewJobRunner(str *store.Store) JobRunner {
+func NewJobRunner(jobExecutor JobExecutor) JobRunner {
 	return &jobRunner{
-		// Unscoped allows the processing of runs that are soft deleted asynchronously
-		store:   str.Unscoped(),
-		workers: make(map[string]chan struct{}),
+		workers:     make(map[string]struct{}),
+		jobExecutor: jobExecutor,
 	}
 }
 
-// Start reinitializes runs and starts the execution of the store's runs.
-func (rm *jobRunner) Start() error {
-	rm.bootMutex.Lock()
-	defer rm.bootMutex.Unlock()
-
-	if rm.started {
-		return errors.New("JobRunner already started")
-	}
-	rm.done = make(chan struct{})
-	rm.started = true
-
-	var starterWg sync.WaitGroup
-	starterWg.Add(1)
-	go rm.demultiplexRuns(&starterWg)
-	starterWg.Wait()
-
-	rm.demultiplexStopperWg.Add(1)
+// Start prepares the job runner for accepting runs to execute.
+func (jr *jobRunner) Start() error {
```

Comment:
The previous job runner had a problem: there was no way for a "runChannel" to be cleaned up. This design is far simpler and no queueing is needed.

Reply:
Queuing as been added, but I have maintained a way for the goroutine to finish safely.